README.md
<!--
 * @Description: 
 * @Version: 2.0
 * @Autor: mayako
 * @Date: 2019-12-23 22:43:34
 * @LastEditors: mayako
 * @LastEditTime: 2021-03-02 17:40:52
-->


CHANGELOG.md
## 0.9.17 (2021-03-04)

### 🌟 新功能
范围|描述|commitId
--|--|--
 global | 模版修改 | [d7669b0](https://github.com/luoxue-victor/commitlint/commit/d7669b0)


 - | 初始化 | [fb8d7f8](https://github.com/luoxue-victor/commitlint/commit/fb8d7f8)


 - | 新版本 | [f2fd007](https://github.com/luoxue-victor/commitlint/commit/f2fd007)




### 🐛 Bug 修复
范围|描述|commitId
--|--|--
 - | hbs格式化问题 | [5a45645](https://github.com/luoxue-victor/commitlint/commit/5a45645)


 - | 修复因为generatorAPI导致缺少换行符的问题 | [82d6f1a](https://github.com/luoxue-victor/commitlint/commit/82d6f1a)


 - | 自动添加空格防止删除模版中的空行 | [ef728eb](https://github.com/luoxue-victor/commitlint/commit/ef728eb)





generator/template/_cz-config.js
/*
 * @Description:
 * @Version: 2.0
 * @Autor: mayako
 * @Date: 2021-01-26 16:33:15
 * @LastEditors: mayako
 * @LastEditTime: 2021-03-04 14:56:02
 */
'use strict'
module.exports = {
  types: [
    { value: 'feat', name: '🌟 新功能:     🌟  一个新功能' },
    { value: 'fix', name: '🐛 Bug 修复:      🐛  一个 bug 修复' },
    { value: 'docs', name: '📝 文档:    📝  只有文档发生改变' },
    {
      value: 'style',
      name:
        '🎨 代码样式:    🎨  对代码含义无影响的改动（空格，格式化，等，非 UI 变动）',
    },
    {
      value: 'refactor',
      name:
        '🔨 代码重构: 🔨  既不是修复 bug 也不是添加新功能的代码更改',
    },
    {
      value: 'perf',
      name: '🚀 性能优化:     🚀  提升性能的代码更改',
    },
    { value: 'test', name: '🔧 测试:     🔧  添加一些缺失的测试或者修正已存在的测试' },
    {
      value: 'chore',
      name:
        '🏠 构建:    🏠  影响构建系统或外部依赖的更改（例如：gulp，npm，webpack）',
    },
    { value: 'revert', name: '🔙 撤销:   🔙  撤销上一次的提交' },
    { value: 'WIP', name: 'WIP:      进行中' },
  ],
  scopes: [
    { name: 'global' },
    { name: 'ui' },
    { name: 'data' },
    { name: 'component' },
    { name: 'unknown' },
  ],
  // it needs to match the value for field type. Eg.: 'fix'
  /*
scopeOverrides: {
fix: [
{name: 'merge'},
{name: 'style'},
{name: 'e2eTest'},
{name: 'unitTest'}
]
},
*/
  // override the messages, defaults are as follows
  messages: {
    type: "选择你提交的信息类型:",
    scope: '\n（非必填）本次提交的改变所影响的范围？',
    // used if allowCustomScopes is true
    customScope: '标示这次的变更:',
    subject: '（必填）写一个简短的变化描述:\n',
    body:
      '（非必填）提供更详细的变更描述. 使用 "|" 断行:\n',
    breaking: '（非必填）列出所有的不兼容变更:\n',
    footer:
      '输入这次涉及的jira号 例如.: #OPENAPI-946, #OPENAPI-947:\n',
    confirmCommit: '你确认要提交么？',
  },
  allowCustomScopes: true,
  allowBreakingChanges: ['feat', 'fix'],
  // limit subject length
  subjectLimit: 100,
}


generator/template/changelog/writer-opts.js
const compareFunc = require('compare-func')
const Q = require('q')
const readFile = Q.denodeify(require('fs').readFile)
const resolve = require('path').resolve

module.exports = Q.all([
  readFile(resolve(__dirname, './templates/template.hbs'), 'utf-8'),
  readFile(resolve(__dirname, './templates/header.hbs'), 'utf-8'),
  readFile(resolve(__dirname, './templates/commit.hbs'), 'utf-8'),
  readFile(resolve(__dirname, './templates/footer.hbs'), 'utf-8'),
]).spread((template, header, commit, footer) => {
  const writerOpts = getWriterOpts()

  writerOpts.mainTemplate = template
  writerOpts.headerPartial = header
  // 经过vue-cli的generator api 会导致缺少最后一个换行，在这补上
  commit = commit + ' '
  writerOpts.commitPartial = commit
  writerOpts.footerPartial = footer

  return writerOpts
})

function getWriterOpts() {
  return {
    transform: (commit, context) => {
      // 新增显示提交者信息字段
      let changelog = context.packageData.changelog
      let gitUserInfo = ''
      let authorName = changelog ? changelog.authorName || false : false
      let authorEmail = changelog ? changelog.authorEmail || false : false
      if (authorName && authorEmail) {
        gitUserInfo = `by: **${commit.authorName}** (${commit.authorEmail})`
      }
      if (authorName && authorEmail === false) {
        gitUserInfo = `by: **${commit.authorName}**`
      }
      if (authorName === false && authorEmail) {
        gitUserInfo = `by: (${commit.authorEmail})`
      }
      commit.gitUserInfo = gitUserInfo
      let discard = false
      const issues = []

      commit.notes.forEach((note) => {
        note.title = '💡 不兼容变更'
        discard = true
      })

      if (commit.type === 'feat') {
        commit.type = '🌟 新功能'
      } else if (commit.type === 'fix') {
        commit.type = '🐛 Bug 修复'
      } else if (commit.type === 'perf') {
        commit.type = '🚀 性能优化'
      } else if (commit.type === 'revert') {
        commit.type = '🔙 撤销'
      } else if (discard) {
        return
      } else if (commit.type === 'docs') {
        commit.type = '📝 文档'
      } else if (commit.type === 'style') {
        commit.type = '🎨 代码样式'
      } else if (commit.type === 'refactor') {
        commit.type = '🔨 代码重构'
      } else if (commit.type === 'test') {
        commit.type = '🔧 测试'
      } else if (commit.type === 'build') {
        commit.type = '🏠 构建系统'
      } else if (commit.type === 'ci') {
        commit.type = '📦 持续集成'
      } else if (commit.type === '杂') {
        commit.type = '🚴 杂'
      }

      if (commit.scope === '*') {
        commit.scope = ''
      }

      if (typeof commit.hash === 'string') {
        commit.hash = commit.hash.substring(0, 7)
      }

      if (typeof commit.subject === 'string') {
        let url = context.repository
          ? `${context.host}/${context.owner}/${context.repository}`
          : context.repoUrl
        if (url) {
          url = `${url}/issues/`
          // Issue URLs.
          commit.subject = commit.subject.replace(/#([0-9]+)/g, (_, issue) => {
            issues.push(issue)
            return `[#${issue}](${url}${issue})`
          })
        }
        if (context.host) {
          // User URLs.
          commit.subject = commit.subject.replace(
            /\B@([a-z0-9](?:-?[a-z0-9]){0,38})/g,
            `[@$1](${context.host}/$1)`
          )
        }
      }

      // remove references that already appear in the subject
      commit.references = commit.references.filter((reference) => {
        if (issues.indexOf(reference.issue) === -1) {
          return true
        }
        return false
      })

      return commit
    },
    groupBy: 'type',
    commitGroupsSort: 'title',
    commitsSort: ['scope', 'subject'],
    noteGroupsSort: 'title',
    notesSort: compareFunc,
  }
}


generator/template/changelog/index.js
/*
 * @Description: 
 * @Version: 2.0
 * @Autor: mayako
 * @Date: 2019-12-23 22:43:34
 * @LastEditors: mayako
 * @LastEditTime: 2021-03-02 16:31:30
 */
'use strict';
const Q = require(`q`);
const conventionalChangelog = require(`./conventional-changelog`);
const parserOpts = require(`./parser-opts`);
const recommendedBumpOpts = require(`./conventional-recommended-bump`);
const writerOpts = require(`./writer-opts`);
// 格式化 git log 信息
const gitRawCommitsOpts = require('./git-raw-commit');
// GitHub/GitLab 信息格式，因为 GitLab 通过 commits 无法访问到具体的记录，但 GitHub 都支持
let context = {
  commit: 'commit'
};

module.exports = Q.all([
  conventionalChangelog,
  parserOpts,
  recommendedBumpOpts,
  writerOpts,
  gitRawCommitsOpts
]).spread(
  (
    conventionalChangelog,
    parserOpts,
    recommendedBumpOpts,
    writerOpts,
    gitRawCommitsOpts
  ) => {
    return {
      context,
      conventionalChangelog,
      parserOpts,
      recommendedBumpOpts,
      writerOpts,
      gitRawCommitsOpts
    };
  }
);
